### PR TITLE
Add Human Pose Estimation for FluidCV

### PR DESCRIPTION
**Pull request by kaiyuyue**
This code is a re-implementation of Simple Baselines for Human Pose Estimation in Fluid.
Its known issue of BN is tracked at [here](https://github.com/PaddlePaddle/Paddle/issues/14580).

### CODE REVIEW

File: fluid/PaddleCV/human_pose_estimation/LICENSE [hunk 1]
```diff
@@ -0,0 +1,203 @@
+Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/LICENSE]: 去掉 LICENSE吧，其他模型都没有加。

File: fluid/PaddleCV/human_pose_estimation/lib/JointsDataset.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) Microsoft
+# Licensed under the MIT License.
+# Written by Bin Xiao (Bin.Xiao@microsoft.com)
+# ------------------------------------------------------------------------------
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/lib/JointsDataset.py]: 这个代码加下说明？ 说下是引用xx的代码，同时把这个Copyright保留？

File: fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+    #     'image': image_file,
+    #     'filename': filename,
+    #     'imgnum': imgnum,
+    #     'joints': joints,
+    #     'joints_vis': joints_vis,
+    #     'center': c,
+    #     'scale': s,
+    #     'rotation': r,
+    #     'score': score
+    # }
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py]: remove the unused code.

File: fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    #     'image': image_file,
+    #     'filename': filename,
+    #     'imgnum': imgnum,
+    #     'joints': joints,
+    #     'joints_vis': joints_vis,
+    #     'center': c,
+    #     'scale': s,
+    #     'rotation': r,
+    #     'score': score
+    # }
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py]: same as above

File: fluid/PaddleCV/human_pose_estimation/README.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+- Flip test is used.
+- We do not hardly search the best model, just use the last saved model to make validation.
+
+## Getting Start
+
+### Prepare Datasets and Pretrained Models
+
+- Following the [instruction](https://github.com/Microsoft/human-pose-estimation.pytorch#data-preparation) to prepare datasets.
+- Download the pretrained ResNet-50 model in PaddlePaddle.Fluid on ImageNet from [Model Zoo](https://github.com/PaddlePaddle/models/tree/develop/fluid/PaddleCV/image_classification#supported-models-and-performances). Put it in the folder `pretrained` under the directory root of this repo.
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/README.md]: 这个给一个明确的下载运行脚本和链接吧，image_classification会增加更多的模型，以及带name的ResNet模型，防止用户不知道要下载哪个

wget http:/xxx

File: fluid/PaddleCV/human_pose_estimation/lib/pose_resnet.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+from __future__ import division
+from __future__ import print_function
+import paddle.fluid as fluid
+
+__all__ = ["ResNet", "ResNet50", "ResNet101", "ResNet152"]
+
+BN_MOMENTUM = 0.1
+
+class ResNet():
+    def __init__(self, layers=50, kps_num=16, test_mode=False):
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/lib/pose_resnet.py:29]: Need comments for these arguments. 

And mabye the `BN_MOMENTUM` also can be put in these arguments.

File: fluid/PaddleCV/human_pose_estimation/train.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+            cv2.circle(masked_image,
+                       (int(preds[i][j][0]), int(preds[i][j][1])),
+                       1, [0, 0, 255], 1)
+
+            width_begin = heatmap_width * (j+1)
+            width_end = heatmap_width * (j+2)
+            grid_image[height_begin:height_end, width_begin:width_end, :] = \
+                masked_image
+            # grid_image[height_begin:height_end, width_begin:width_end, :] = \
+            #     colored_heatmap*0.7 + resized_image*0.3
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/train.py]: Maybe remove these un-used code?

File: fluid/PaddleCV/human_pose_estimation/utils/transforms.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) Microsoft
+# Licensed under the MIT License.
+# Written by Bin Xiao (Bin.Xiao@microsoft.com)
+# ------------------------------------------------------------------------------
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/utils/transforms.py:22]: same as above.  How other framework handle the reference like this?

File: fluid/PaddleCV/human_pose_estimation/val.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    params = {}
+    params["total_images"] = args.total_images
+    params["lr"] = args.lr
+    params["num_epochs"] = args.num_epochs
+    params["learning_strategy"] = {}
+    params["learning_strategy"]["batch_size"] = args.batch_size
+    params["learning_strategy"]["name"] = args.lr_strategy
+
+    if args.with_mem_opt:
+        fluid.memory_optimize(fluid.default_main_program())
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/val.py]: Betther to use [`skip_opt_set`](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/transpiler/memory_optimization_transpiler.py#L457) to skip the `loss`, `output` and `target`, instead of  setting

```
    loss.persistable = True
    output.persistable = True
    target.persistable = True
```


Same as other files.

File: fluid/PaddleCV/human_pose_estimation/val.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+
+    place = fluid.CUDAPlace(0) if args.use_gpu else fluid.CPUPlace()
+    exe = fluid.Executor(place)
+    exe.run(fluid.default_startup_program())
+
+    args.pretrained_model = './pretrained/resnet_50/115'
+    if args.pretrained_model:
+        def if_exist(var):
+            exist_flag = os.path.exists(os.path.join(args.pretrained_model, var.name))
+            # print(var.name, exist_flag)
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/val.py]: remove this line

File: fluid/PaddleCV/human_pose_estimation/LICENSE (hunk 1 above)
Comment (KaiyuYue) [reply fluid/PaddleCV/human_pose_estimation/LICENSE]: done.

File: fluid/PaddleCV/human_pose_estimation/lib/JointsDataset.py (hunk 2 above)
Comment (KaiyuYue) [reply fluid/PaddleCV/human_pose_estimation/lib/JointsDataset.py]: Done. Please check [this part](https://github.com/PaddlePaddle/models/pull/1474/files#diff-a54e5b6b371603cd4adfdd3bd49006c1R1) out.

File: fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py (hunk 3 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/README.md (hunk 5 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/README.md]: Done.

File: fluid/PaddleCV/human_pose_estimation/lib/pose_resnet.py (hunk 6 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/lib/pose_resnet.py:29]: Done.

File: fluid/PaddleCV/human_pose_estimation/val.py (hunk 9 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/val.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+        # sanitize bboxes
+        valid_objs = []
+        for obj in objs:
+            x, y, w, h = obj['bbox']
+            x1 = np.max((0, x))
+            y1 = np.max((0, y))
+            x2 = np.min((width - 1, x1 + np.max((0, w - 1))))
+            y2 = np.min((height - 1, y1 + np.max((0, h - 1))))
+            if obj['area'] > 0 and x2 >= x1 and y2 >= y1:
+                # obj['clean_bbox'] = [x1, y1, x2, y2]
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py]: remove this line ?

File: fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+            img_y = max(0, ul[1]), min(br[1], HEATMAP_SIZE[1])
+
+            v = target_weight[joint_id]
+            if v > 0.5:
+                target[joint_id][img_y[0]:img_y[1], img_x[0]:img_x[1]] = \
+                    g[g_y[0]:g_y[1], g_x[0]:g_x[1]]
+
+    return target, target_weight
+
+def data_augmentation(sample, is_train):
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py:64]: The data aug between mpii and coco is different?

Is there any same fucntion in mpii_reader.py and coco_reader.py? If have, please put them in an common file

File: fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        if joints_vis[i, 0] > 0.0:
+            joints[i, 0:2] = affine_transform(joints[i, 0:2], trans)
+
+    # Numpy target
+    target, target_weight = _generate_target(joints, joints_vis)
+
+    if DEBUG:
+        _visualize(filename, data_numpy, input.copy(), joints, target)
+
+    # input_no_norm = input.copy()
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py]: Remove the unused code.

File: fluid/PaddleCV/human_pose_estimation/test.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    # image and target
+    image = layers.data(name='image', shape=[3, IMAGE_SIZE[1], IMAGE_SIZE[0]], dtype='float32')
+    file_id = layers.data(name='file_id', shape=[1,], dtype='int')
+
+    # build model
+    model = pose_resnet.ResNet(layers=50, kps_num=args.kp_dim, test_mode=True)
+
+    # output
+    output = model.net(input=image, target=None, target_weight=None)
+    output.persistable = True
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/test.py]: remove this line

File: fluid/PaddleCV/human_pose_estimation/test.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+                    feed=feeder.feed(data_fliped))
+
+            # flip back
+            output_flipped = flip_back(output_flipped, FLIP_PAIRS)
+
+            # feature is not aligned, shift flipped heatmap for higher accuracy
+            if args.shift_heatmap:
+                output_flipped[:, :, :, 1:] = \
+                        output_flipped.copy()[:, :, :, 0:-1]
+                # output_flipped[:, :, :, 0] = 0
```
Comment (qingqing01) [inline fluid/PaddleCV/human_pose_estimation/test.py]: remove the un-used code, also in other files.  
若干debug的注释掉的code可以留着，大多没用的code请去掉吧。

File: fluid/PaddleCV/human_pose_estimation/LICENSE (hunk 1 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/LICENSE]: Done.

File: fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py (hunk 4 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/train.py (hunk 7 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/train.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/utils/transforms.py (hunk 8 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/utils/transforms.py:22]: Done.

File: fluid/PaddleCV/human_pose_estimation/val.py (hunk 10 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/val.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py (hunk 11 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/lib/coco_reader.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py (hunk 13 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/lib/mpii_reader.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/test.py (hunk 14 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/test.py]: Done.

File: fluid/PaddleCV/human_pose_estimation/test.py (hunk 15 above)
Comment (kaiyuyue) [reply fluid/PaddleCV/human_pose_estimation/test.py]: Done.

**Review by qingqing01 — APPROVED** [review]
